### PR TITLE
Backup restore to address special characters

### PR DESCRIPTION
**Pull request by tangp3**
Made a system wide change to support special chars in database name, schema name, and table name.

This commit fixes the quotation and escaption for the
whole backup and restore, and resolves related issues which
block us from supporting special chars.

List of remarkable changes:

Removed gpcrondump shell script.
Renamed gpcrondump.py to gpcrondump

Added -S option in gpdbrestore for schema level restore instead of using wildcard '*'.
Added few methods to parse and extract databases, schemas, tables, and others ddl from plain dump file to support special chars
Added some behave tests under gppylib/test/behave/mgmt_utils/

Fixed gpcrondump deleting user specified schema file
Fixed gpdbrestore not validating non existed tables correctly.
Fixed the pg_get_partition_def() and pg_get_partition_template_def() function to
      dump DDL statement properly with quotation and escaping.
Fixed ddboost restore and Post data restore not forming filter command correctly

Updated -x option in gpcrondump to not use comma separated specification.
Updated -T option in gpdbrestore to not use comma separated specification.
Updated the pre and post restore filter to parse special chars within schema and table names
Updated some unit tests, behave tests and the help docs

The commit resolves special chars for all general dump and restore options except few chars:
tab (\t), exclamation mark(!), comma(,), dot (.), newline(\n), as dump and restore uses plain file,
and needs separators,  we will handle them separately in follow up commits.

Authors: Pengcheng Tang, Chumki Roy

### CODE REVIEW

**Comment by hlinnaka** [comment]
On 22/02/16 10:41, Pengcheng Tang wrote:
> Made a system wide change to support special chars in database name, schema name, and table name.
>
> This commit fixes the quotation and escaption for the
> whole backup and restore, and resolves related issues which
> block us from supporting special chars.

Hooray, thanks for cleaning this up!

- Heikki

File: gpMgmt/bin/gpdbrestore [hunk 1]
```diff
@@ -612,8 +623,10 @@ def create_parser():
                      help="Dispatches work to segment hosts in batches of specified size [default: %s]" % DEFAULT_NUM_WORKERS)
     addTo.add_option('-G', dest='restore_global', action="callback", callback=restore_global_callback,
                      help="Restore global objects dump file if found in target dumpset. The global objects file is secured via the gpcrondump -G options, and has a filename of the form gp_global_1_1_<timestamp>. Specify \"-G only\" to only restore the global objects dump file or \"-G include\" to restore globals objects along with a normal restore. Defaults to \"include\" if neither argument is provided.")
-    addTo.add_option('-T', dest='restore_tables', metavar='<schema.tablename>',
-                     help="Restore a single table from a backup set. For multiple tables, supply a comma separated list of schema.tablenames. Note, table schema must exist in target database.")
+    addTo.add_option('-S', action='append', dest='restore_schemas', default=[], metavar='<schema name>',
```
Comment (hlinnaka) [inline gpMgmt/bin/gpdbrestore]: Is -S a new feature? It seems unrelated to fixing the escaping.

File: gpMgmt/bin/gppylib/operations/backup_utils.py [hunk 2]
```diff
@@ -180,17 +189,58 @@ def check_cdatabase_exists(dbname, report_file, dump_prefix, ddboost=False, netb
...
+    """
+    Line format: CREATE DATABASE "DBNAME" WITH TEMPLATE = template0 ENCODING = 'UTF8' OWNER = gpadmin;
+
+    To get the dbname:
+    substring between the ending index of the first statement: CREATE DATABASE and the starting index
+    of WITH TEMPLATE whichever is not inside any double quotes, based on the fact that double quote
+    inside any name will be escaped by extra double quote, so there's always only one WITH TEMPLATE not
+    inside any doubles, means its previous and post string should have only even number of double
+    quotes.
+    Note: OWER name can also have special characters with double quote.
```
Comment (hlinnaka) [inline gpMgmt/bin/gppylib/operations/backup_utils.py]: typo: /s/OWER/OWNER/

File: gpMgmt/bin/gppylib/operations/restore.py [hunk 3]
```diff
@@ -54,17 +58,21 @@ def update_ao_statistics(master_port, dbname, restored_tables):
...
     conn = None
     counter = 1
+    restored_ao_tables = set()
     try:
         results = execute_sql(qry, master_port, dbname)
-        ao_tables = ['%s.%s' % (sch, tbl) for (tbl, sch) in results]
-        restored_ao_tables = set(ao_tables).intersection(restored_tables)
-        if not restored_ao_tables:
+        for (tbl, sch) in results:
+            tblname = '%s.%s' % (sch, tbl)
```
Comment (hlinnaka) [inline gpMgmt/bin/gppylib/operations/restore.py]: Is this missing escaping?

File: gpMgmt/bin/gpdbrestore (hunk 1 above)
Comment (tangp3) [inline gpMgmt/bin/gpdbrestore]: -S is a new option to replace the old wildcard feature for schema level restore (by specifying schemaname.*  ), it's closely related to this fix as we want to support the star char '*'  as part of schema name, there are few other fixes related to this commit, the rational is to make it possible for addressing special chars.

File: gpMgmt/bin/gppylib/operations/backup_utils.py (hunk 2 above)
Comment (tangp3) [inline gpMgmt/bin/gppylib/operations/backup_utils.py]: thanks, to replace.

File: gpMgmt/bin/gppylib/operations/restore.py (hunk 3 above)
Comment (tangp3) [inline gpMgmt/bin/gppylib/operations/restore.py]: Here it is right, the user provided tables (restored_tables, here) are with all original names without escaping and quotation, we handle the escaping and quotation only when necessary within each components which use them.

Comment (hlinnaka) [inline gpMgmt/bin/gppylib/operations/restore.py]: On 22/02/16 18:25, Pengcheng Tang wrote:
>>       try:
>>           results = execute_sql(qry, master_port, dbname)
>> -        ao_tables = ['%s.%s' % (sch, tbl) for (tbl, sch) in results]
>> -        restored_ao_tables = set(ao_tables).intersection(restored_tables)
>> -        if not restored_ao_tables:
>> +        for (tbl, sch) in results:
>> +            tblname = '%s.%s' % (sch, tbl)
>
> Here it is right, the user provided tables (restored_tables, here)
> are with all original names without escaping and quotation, we handle
> the escaping and quotation only when necessary within each components
> which use them.

If the table or schema name contains a dot (.), how does that work?

- Heikki

File: gpMgmt/bin/gpdbrestore (hunk 1 above)
Comment (hlinnaka) [inline gpMgmt/bin/gpdbrestore]: On 22/02/16 18:16, Pengcheng Tang wrote:
>> @@ -612,8 +623,10 @@ def create_parser():
>>                        help="Dispatches work to segment hosts in batches of specified size [default: %s]" % DEFAULT_NUM_WORKERS)
>>       addTo.add_option('-G', dest='restore_global', action="callback", callback=restore_global_callback,
>>                        help="Restore global objects dump file if found in target dumpset. The global objects file is secured via the gpcrondump -G options, and has a filename of the form gp_global_1_1_<timestamp>. Specify \"-G only\" to only restore the global objects dump file or \"-G include\" to restore globals objects along with a normal restore. Defaults to \"include\" if neither argument is provided.")
>> -    addTo.add_option('-T', dest='restore_tables', metavar='<schema.tablename>',
>> -                     help="Restore a single table from a backup set. For multiple tables, supply a comma separated list of schema.tablenames. Note, table schema must exist in target database.")
>> +    addTo.add_option('-S', action='append', dest='restore_schemas', default=[], metavar='<schema name>',
>
> -S is a new option to replace the old wildcard feature for schema
> level restore (by specifying schemaname.*  ), it's closely related to
> this fix as we want to support the star char '*'  as part of schema
> name, there are few other fixes related to this commit, the rational
> is to make it possible for addressing special chars.

Ok. Do we still support the old wildcard feature, too? Would it make 
sense to split the -S feature into a separate patch?

- Heikki

File: gpMgmt/bin/gppylib/operations/restore.py (hunk 3 above)
Comment (tangp3) [inline gpMgmt/bin/gppylib/operations/restore.py]: We attempted to add support for few other chars like ! (messing up with shell cmd extension, must use single quote all way down from shell to python to C) and dot(.),  which turned out we need to change behavior of few other components as well.

for instance, for dot (.)
partly due to the fact that we are using plain file to specify

 <schema>.<table> in input file,  (using formatted file like json or xml may be better option in future)

and similarly gpcrondump uses format of: 

timestamp: <schema>.<table>, <schema>.<table> for gp_dump_table_list file.

I initially attempted the approach of quotation on dot(.), but it turns out to be complex and error prone when distinguishing whether or not the quotes belongs to a schema/table name, therefore, I pushed it back to next story in backlog. 

To the end, I'd say the current implementation based on plain file does not lend us the easy support.

File: gpMgmt/bin/gpdbrestore (hunk 1 above)
Comment (tangp3) [inline gpMgmt/bin/gpdbrestore]: Understand it'd be easier to maintain as a separate patch, will put into a separate commit.

We don't support the old wildcard one as we treat all chars as part of relname.

This option was introduced mainly to provide support of wildcard chars, so I put it as part
of the commit to give people the rational.

File: gpMgmt/bin/gppylib/operations/restore.py (hunk 3 above)
Comment (tangp3) [inline gpMgmt/bin/gppylib/operations/restore.py]: "If the table or schema name contains a dot (.), how does that work?"

We will work on supporting dot(.) tracked in backlog, currently can do a check_funny_chars_in_names() to prevent this.

**Comment by tangp3** [comment]
Pushed to master, Thanks for reviewing.
